### DISCUSSION TITLE
Maintaining a version string in a project and using it for creating releases

### DISCUSSION THREAD

**Discussion by SayakMukhopadhyay** [community/community#26905] [Actions]
<p>Hi</p>
<p>I was looking into automating our release process with Actions and I am in a fix as to how I can best achieve a clean methodology. We want to make releases manually so something like <a href="https://github.com/semantic-release/semantic-release" rel="nofollow noopener">Semantic Release</a> is not going to be a good fit. As such I decided that we can create an action which creates a release and gets triggered when we want a release. The question was how to trigger this action.</p>
<p>For that, I went with creating a <code>VERSION</code> file which just has the version string. The release action gets triggered only when this file changes which works well for now. But, creation of this file means that the version number is duplicated in this file and in the project’s native manifest file (like package.json, build.gradle etc). Here’s a few more ideas I have:</p>
<ol>
<li>I can remove the version information in such manifest files but there might be some framework which screams and throws a fit if I do so.</li>
<li>I can put a fake value in the version field of these manifest files but this seems to uncomfortable hacky</li>
<li>I can drop the <code>VERSION</code> file idea and trigger the release action on change to the manifest file itself. Trouble is, that I can’t trigger the release everytime the manifest changes since it can change without the version information being changed like updation of dependencies. I would need to specifically check if the version field has changed. I would need to handle edge cases like the line number of the version field getting relocated elsewhere for which I would need to check against the same file with the previous version, etc.</li>
</ol>
<p>The 3rd solution seems to me to be the most flexible but it needs some effort in getting right. I was looking for opinions and insights as to which of these ideas is the best or if there is an even better idea.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/sayakmukhopadhyay">@sayakmukhopadhyay</a> ,</p>
<blockquote>
<p>As such I decided that we can create an action which creates a release and gets triggered when we want a release.</p>
</blockquote>
<p>In fact, we have an official action provided by GitHub to create release (<a href="https://github.com/actions/create-release" rel="nofollow noopener"><strong>actions/create-release</strong></a>).</p>
<br>
<blockquote>
<p>The question was how to trigger this action.</p>
</blockquote>
<p>When creating a release, we need to provide a tag which generally is used as the version of the new release.<br>
To trigger the action to create release in the workflow, here I have two ways as reference:</p>
<ol>
<li>
<p>When a specified tag is pushed to the repository, trigger the workflow.</p>
<pre><code class="lang-yaml">name: Release
on:
  push:
    tags:
      - '&lt;tags_filter_patterns&gt;'
# Only the tags match the filter patterns can trigger the workflow.
# See: https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestbranchestags

jobs:
  job1:
    name: Create release
    runs-on: ubuntu-latest
    steps:
      - name: Create Release
        id: create_release
        uses: actions/create-release@v1
        with:
          tag_name: ${{ github.ref }}
          . . .
</code></pre>
<p>This way generally is applied when you manually create a tag and push it to the repository.</p>
</li>
<li>
<p>If you want to automatically create a tag from the commit that you want to release, you also need to provide the name of the new tag and the commit SHA or the branch name on the action.</p>
<pre><code class="lang-yaml">name: Release
on:
  push:
    branches:
      - '&lt;branches_filter_patterns&gt;'

jobs:
  job1:
    name: Create release
    runs-on: ubuntu-latest
    steps:
      - name: Create Release
        id: create_release
        uses: actions/create-release@v1
        with:
          tag_name: &lt;tag_name&gt;
          commitish: ${{ github.sha }}
          . . .
</code></pre>
<p>In this way, you may need to use a “Build Number Generator” action to generate the auto-incrementing build number used for setting the tag name (release version number).<br>
Related actions from the GitHub Marketplace:</p>
<ul>
<li><a href="https://github.com/marketplace/actions/build-number" rel="nofollow noopener"><strong>Build Number</strong></a></li>
<li><a href="https://github.com/marketplace/actions/build-number-generator" rel="nofollow noopener"><strong>Build Number Generator</strong></a></li>
<li><a href="https://github.com/marketplace/actions/nuget-build-number-generator" rel="nofollow noopener"><strong>Nuget Build Number Generator</strong></a></li>
</ul>
<p>In addition, when you commit and push the changes you want to release, you can add a keyword in the commit message so that the workflow can know this commit need to be released.<br>
For example,</p>
<ul>
<li>
<p>set the commit message starts with or contains the keyword “<strong>[release]</strong>”.<br>
<img src="https://user-images.githubusercontent.com/46307996/181100539-07fe19ce-3fee-4f90-b007-58db2bf777d1.png" alt="commit_message" data-base62-sha1="1NnEUme1wYNCNdBXaRM2468cVZ9" width="412" height="153"></p>
</li>
<li>
<p>In the workflow, use the if statement to trigger the action to create release when the specified keyword contained in the commit message.</p>
<pre><code class="lang-yaml">- name: Create Release
  id: create_release
  if: ${{ contains(github.event.head_commit.message, '[release]') }}
  uses: actions/create-release@v1
  with:
    tag_name: &lt;tag_name&gt;
    commitish: ${{ github.sha }}
    . . .
</code></pre>
</li>
</ul>
</li>
</ol>

**Comment by SayakMukhopadhyay** [comment]
<p>Hi <a class="mention" href="/brightran">@brightran</a>, thanks for the response. I didn’t mention it but I am already using the create-release action to create a release.</p>
<blockquote>
<p>As such I decided that we can create an action which creates a release and gets triggered when we want a release.</p>
</blockquote>
<p>I meant that I created a workflow, not an action. Sorry for confusing the terms :pray:.</p>
<p>But anyway, this is how my workflow looks right now</p>
<pre><code class="lang-auto">name: Release
on:
  push:
    branches:
      - master
    paths:
      - 'VERSION'

jobs:
  docker-build:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Set Version ENV
        run: |
          echo '::set-env name=VERSION::'$(cat VERSION)

      - name: Create Release
        uses: actions/create-release@v1
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        with:
          tag_name: ${{ env.VERSION }}
          release_name: ${{ env.VERSION }}
</code></pre>
<p>The <code>VERSION</code> file contains a string like <code>1.2.3</code>. So, when I change the version file to something like <code>1.3.0</code>, and push the commit, this workflow is triggered and a new release is created.</p>
<p>My concern was that everytime I change this file, I also need to change the <code>package.json</code> file’s version field. This can give rise to inconsistencies in the long run which I want to avoid.</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/sayakmukhopadhyay">@sayakmukhopadhyay</a> ,</p>
<p>You can try like as this:<br>
In the workflow, use the ‘<a href="https://www.geeksforgeeks.org/sed-command-in-linux-unix-with-examples/" rel="nofollow noopener"><strong>sed</strong></a>’ command to replace the variable name with the actual version number in the ‘<strong>package.json</strong>’.</p>
<pre><code class="lang-yaml">    steps:
      - name: Checkout code
        uses: actions/checkout@v2

      - name: Set Version ENV
        run: echo "::set-env name=VERSION::'$(cat VERSION)"

      - name: Update version in 'package.json' and push changes
        run: |
          old_str="\"version\": \"1.2.3\""
          new_str="\"version\": \"1.3.0\""
          sed -i 's/$old_str/$new_str/g' package.json

          git config --global user.name "you username"
          git config --global user.email "you email"
          git add package.json
          git commit -m "Update version"
          git push

      - name: Create Release
        uses: actions/create-release@v1
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        with:
          tag_name: ${{ env.VERSION }}
          release_name: ${{ env.VERSION }}
          commitish: master
</code></pre>
<p>In this way, you only need to changes the version number in the VERSION file to trigger the workflow, then the workflow will execute the ‘<strong>sed</strong>’ command to automatically changes the version in the ‘<strong>package.json</strong>’.</p>

**Comment by SayakMukhopadhyay** [comment]
<p>Nice, this is a great idea. Right now, I have implemented another idea which doesn’t require the workflow to make a commit but relies on the reviewer. I have added a workflow job which parses the <code>package.json</code> and the <code>VERSION</code> file and tests if both of them have the same value. If not, this is reflected as a failed test when someone creates a PR. Thus a reviewer can ensure that the PR author fixes this.</p>
<pre><code class="lang-auto">check-version:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v2
        with:
          persist-credentials: false

      - name: Generate version strings
        run: |
          echo '::set-env name=VERSION_FILE::'$(cat VERSION)
          echo '::set-env name=PACKAGE_FILE::'$(sed -e 's/^"//' -e 's/"$//' &lt;&lt;&lt; $(jq .version package.json))
          echo '::set-env name=PACKAGE_LOCK_FILE::'$(sed -e 's/^"//' -e 's/"$//' &lt;&lt;&lt; $(jq .version package-lock.json))

      - name: Verify version strings
        run: |
          [[ "$VERSION_FILE" == "$PACKAGE_FILE" &amp;&amp; "$PACKAGE_FILE" == "$PACKAGE_LOCK_FILE" ]]
</code></pre>
<p>Hopefully this alternate idea can also help some people.</p>
